Add support of h3ulcb and salvator xs m3 arm64 boards

xakep-amatop opened the pull request:
Add generic Renesas functions to CPG for getting/setting rates of clocks. Add associated tables of clocks and API for working with these tables.

Add MMIO mapping for PFC, UART and CPG Renesas drivers in order to avoid mappings inside `mmu_regions.c` file(s).

Added files for supporting  ARM64 Renesas `r8a77951`/`r8a77961` SoCs:
    * add dtsi for `r8a77951`/`r8a77961` soc;
    * files for support PFC;

Add basic support of `rcar_salvator_xs_m3` and `rcar_h3ulcb_ca57` ARM64 boards.

xakep-amatop committed:
boards: arm64: add Renesas H3ULCB CA57 board support

Add basic functionality for supporting h3ulcb board:
  * add documentation for h3ulcb board;
  * add pinctrl dtsi for scif driver;
  * add dts, yaml and configuration files.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
soc: arm64: add Renesas Rcar Gen3 SoC support

Add files for supporting arm64 Renesas r8a77951 SoC.
Add config option CPU_CORTEX_A57.

Enable build of clock_control_r8a7795_cpg_mssr.c for
a new ARM64 SoC R8A77951.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: uart_rcar: do MMIO mapping inside driver

Add MMIO mapping for UART Renesas driver in order to avoid
mappings inside mmu_regions.c file(s).

There are a lot of changes inside SCIF driver inside this
commit, because reg addr and size may be stored in RAM or
ROM and appropriately in different driver structures data
or cfg, and, because the name of field reg base is changed.

Note: it is common approach according to Zephyr documentation.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: pfc_rcar: do MMIO mapping inside driver

Add MMIO mapping for PFC Renesas driver in order to avoid
mappings inside mmu_regions.c file.

Add a new system init function pfc_rcar_driver_init to PFC
Renesas driver for invoking a memory mapping macro.
Note: PFC Renesas driver doesn't use Zephyr Device Model.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: add generic Renesas functions for get/set rate

Add associated tables of clocks and API for working with these tables,
from this moment the relationship between clocks and their divider are
built. After set rate of some Core clock, driver has to update all
in/out rates of all childrens recursively. During get/set rate calls
if out rate is unknown, we try to get parent in/out rates and its
divider, in case when parent doesn't have valid in/out rates we get
parent of parent and so on until we get parent with a valid in or out
rates.

Add generic Renesas functions for get/set rate of CPG.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: redesign cpg driver for r8a7795

Add associative tables for core and mod clocks.

Add possibility to enable/disable any of core clocks which presents
in the associative table.

Add handler for setting rate to driver and use generic rcar cpg
function for getting rate.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: delete irq lock/unlock from mstp_clock_endisable

Delete IRQ lock/unlock calls from 'rcar_cpg_mstp_clock_endisable',
because 'rcar_cpg_mstp_clock_endisable' function is always called
under spin lock.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: do MMIO mapping inside driver

Add MMIO mapping for Renesas CPG driver in order to avoid
mappings inside mmu_regions.c file. Remove MMU region for
Renesas CPG driver.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
drivers: rcar_cpg: move cpg log module declare to common rcar cpg file

Move cpg log module declare to common rcar cpg file. If the module
consists of multiple files, then LOG_MODULE_REGISTER should appear
in exactly one of them. Each other file should use LOG_MODULE_DECLARE
to declare its membership in the module.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

aaillet commented:
Hi @xakep-amatop ,
Before reviewing this work, I would like to know which use cases you are targeting by supporting Cortex-A57 core of these boards in Zephyr.
These boards are equipped with real-time purpose Cortex-R7 cores which we are already supporting (only H3 based boards atm).
A57s and R7 cores may even be able to communicate together through shared msg queues (work is in progress).

I would like to know your though (and even other community members though) about this.

In the case we want to merge this work, I think it would be nice to unify as much as possible our work from the beginning to create a clean working tree containing both arm and arm64 part of these SoCs support.

lorc commented:
Hi @aaillet,

Let me jump in with some clarifications.

> Before reviewing this work, I would like to know which use cases you are targeting by supporting Cortex-A57 core of these boards in Zephyr.

We are working on virtualization on Rcar Gen3/Gen4 platform. If you are interested, you can check out our repo: https://github.com/xen-troops/meta-xt-prod-devel-rcar 

Here is some background: we are running Xen Hypervisor. This is Type-1 hypervisor, which means that it boots before any OS. After boot Xen automatically creates very first virtual machine, which is known as Dom0. This is a bit special VM, designated as  "control" or "privileged" domain, because it is capable of controlling hypervisor and creating other VMs (or "domains", as Xen call them). In most cases Dom0 is a typical Linux distribution with some additional Xen-related tools. This is perfectly fine for generic server use cases, but is not suitable for automotive ones. 

We are planning Xen to be safety-certified, but we need safety-certified Dom0 as well, because Dom0 can control hypervisor behaviour. As it is almost impossible to certify Linux kernel, we need something more specialized and lightweight, which is Zephyr, of course :)

Actually, we already have "Zephyr as Dom0" solution. It is not ready for production yet, but if anyone want to experiment with it, we can provide links to the repos. We are planning to upstream all relevant changes and this PR is the first in series.

So yeah, to run Zephyr as Xen's Dom0, we need Zephyr to be able to run on CortexA cores on RCar Gen3 first. This is why you are seeing this PR.

SgrrZhf commented on boards/arm64/renesas_rcar/gen3/rcar_salvator_xs_m3_2x4g/doc/rcar_salvator_xs_m3_2x4g.rst:
```diff
@@ -0,0 +1,69 @@
+
+Supported Features
+==================
+The Renesas rcar_salvator_xs_m3_2x4g board configuration supports the following
```
Missing a blank line?

https://builds.zephyrproject.io/zephyr/pr/57172/docs/boards/arm64/renesas_rcar/gen3/rcar_salvator_xs_m3_2x4g/doc/rcar_salvator_xs_m3_2x4g.html

aaillet commented on boards/arm64/renesas_rcar/Kconfig.board:
boards file tree is usually not using brand/family folders.
Please get boards folder to the same tree level = boards/arm64/rcar_h3ulcb_ca57

aaillet commented on dts/arm64/renesas/gen3_common/r8a7795_r8a7796_common.dtsi:
Please use same file tree as for arm32 part of renesas socs for clarity purpose.
dts/arm64/renesas/gen3 with
  - rcar_gen3_ca57.dtsi (common)
  - r8a77951.dtsi (r8a77950 SoC is deprecated, no need to support it)
  - r8a7796.dtsi (M3-W + M3-N)

aaillet commented on soc/arm64/renesas_rcar/gen3/Kconfig.defconfig.r8a7795:
please rename to Kconfig.defconfig.r8a77951, we dont support deprecated r8a77950 SoC

aaillet commented on soc/arm64/renesas_rcar/gen3/pfc_r8a7795.c:
please rename to soc/arm64/renesas_rcar/gen3/pfc_r8a77951.c

aaillet commented on boards/arm64/renesas_rcar/gen3/common_salvator_xs_h3ulcb.dts:
It's currently not common to deploy common files in the boards file tree, please put back these fields in each board dts.

aaillet reviewed (dismissed):
Hi, I did a first short review on file tree, please edit it to match what have been done on CR7 side and common Zephyr practice.

Concerning using MMIO mapping in several drivers, are these changes mandatory for the A57 to use these drivers ? If not, please remove these commits from the PR and do another dedicated PR to use MMIO mapping on all r_car drivers !

We should check if we can use one and only `pfc_r8a77951.c` and so on.

Also, for the Zephyr bot to be able to assign the right people when creating Pull requests, separate label should be created for Renesas R-Car arm and Renesas R-Car arm64. Maybe you can add a commit editing `MAINTAINERS.yml`.

I haven't had time to review the rest of the changes yet and I won't be around for two weeks, sorry for that, I'll resume the review then.

xakep-amatop committed:
soc: arm64: add support of r8a77961

Add support of r8a77961 SoC to gen3 series.
Create a dtsi file with a common part for both r8a77951 and r8a77961.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
soc: arm64: add PFC files to Renesas r8a77961 Gen3 SoC

Add Pin Function Controller tables of registers and their bits
for ARM64 Renesas R-Car family. With this changes we can use
Renesas PFC driver for configuring bias and driving capabilities.

Add only needed driver strength and bias pins to PFC,
e.g. SDx and UART TX/RX pins.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop committed:
boards: arm64: add support of Salvator XS M3 board

Add support of 'rcar_salvator_xs_m3' board: minimal dts
and configuration.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

xakep-amatop commented:
> Hi, I did a first short review on file tree, please edit it to match what have been done on CR7 side and common Zephyr practice.

thanks, fixed!

> Concerning using MMIO mapping in several drivers, are these changes mandatory for the A57 to use these drivers ? If not, please remove these commits from the PR and do another dedicated PR to use MMIO mapping on all r_car drivers !

most of changes inside drivers are necessary in order to get working UART and to test the code, so I've put drivers related changes to this PR

> We should check if we can use one and only pfc_r8a77951.c and so on.

do you mean to share this file between arm32 & arm64 SoCs or only between r8a77951 & r8a7796 for arm64?

> Also, for the Zephyr bot to be able to assign the right people when creating Pull requests, separate label should be created for Renesas R-Car arm and Renesas R-Car arm64. Maybe you can add a commit editing MAINTAINERS.yml.

Do want to become a maintainer for Renesas R-Car arm64 too? I'm afraid that I have not enough experience for this role. Maybe you know somebody who wants and can be a maintainer?

xakep-amatop committed:
soc: arm64: add PFC files to Renesas r8a77951 Gen3 SoC

Add Pin Function Controller tables of registers and their bits
for ARM64 Renesas R-Car family. With this changes we can use
Renesas PFC driver for configuring bias and driving capabilities.

Note: some of files copy-pasted from Renesas Arm 32
      SoC directory and this commit is a temporary
      solution, we need to rework it in order to use
      the same source files for both architectures.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>

harunobukurokawa reviewed (commented):
Hi, 
Please confirm the revision of M3. These are difference between "r8a77960" and "r8a77961".

xakep-amatop commented on dts/arm64/renesas/r8a7796.dtsi:
I've used Salvator XS M3 ver 3.0 for tests

according to this page https://elinux.org/R-Car/Boards/Salvator-XS:
both r8a77960 and r8a77961 for M3:
```
 * DTB: arch/arm64/boot/dts/renesas/r8a77960-salvator-xs.dtb (for R-Car M3-W)
 * DTB: arch/arm64/boot/dts/renesas/r8a77961-salvator-xs.dtb (for R-Car M3-W+)
 * DTB: arch/arm64/boot/dts/renesas/r8a77965-salvator-xs.dtb (for R-Car M3-N)
```

u-boot on test board reports the next info:
```
CPU: Renesas Electronics R8A7796 rev 3.0
Model: Renesas Salvator-X board based on r8a77960
DRAM:  7.9 GiB
```

but sticker on board says that this is XS board and BL2 approves this info:
```
NOTICE:  BL2: R-Car Gen3 Initial Program Loader(CA57) Rev.3.0.1
NOTICE:  BL2: PRR is R-Car M3 Ver.3.0
NOTICE:  BL2: Board is Salvator-XS Rev.1.0
```

harunobukurokawa replied:
Thank you for your information. 
It is "r8a77961" because you use M3 ver 3.0. The description from u-boot is not correct.

xakep-amatop commented:
fixed naming of files used for documentation

xakep-amatop commented on dts/arm64/renesas/r8a7796.dtsi:
```diff
@@ -0,0 +1,28 @@
+/*
+ * Device Tree Source for the R-Car H3/M3 (R8A7796) SoC
```
used `r8a77961` name instead of `r8a7796`/`r8a77960`

carlescufi commented:
@xakep-amatop please rebase
@gmarull @aaillet please re-review

carlescufi commented:
> > @xakep-amatop please rebase @gmarull @aaillet please re-review
> 
> @carlescufi thank you, I've fixed merge conflicts!

@xakep-amatop thanks, please take a look at the failing compliance check.

xakep-amatop commented:
> > > @xakep-amatop please rebase @gmarull @aaillet please re-review
> > 
> > 
> > @carlescufi thank you, I've fixed merge conflicts!
> 
> @xakep-amatop thanks, please take a look at the failing compliance check.

@carlescufi as far as I understand using of `typedef`s for `pinctrl_soc_pin_t` it is a common approach, look at how it was made for other SoCs (part of output):
```
###### grep -rnI pinctrl_soc_pin_t

arm64/nxp_imx/mimx9/pinctrl_soc.h:52:typedef struct pinctrl_soc_pin pinctrl_soc_pin_t;
arm64/nxp_imx/mimx8m/pinctrl_soc.h:52:typedef struct pinctrl_soc_pin pinctrl_soc_pin_t;
arm64/renesas_rcar/gen4/pinctrl_soc.h:41:} pinctrl_soc_pin_t;
riscv/esp32c3/pinctrl_soc.h:28:} pinctrl_soc_pin_t;
riscv/openisa_rv32m1/pinctrl_soc.h:23:typedef uint32_t pinctrl_soc_pin_t;
riscv/riscv-privilege/telink_b91/pinctrl_soc.h:17:typedef uint32_t pinctrl_soc_pin_t;
riscv/riscv-privilege/sifive-freedom/pinctrl_soc.h:12:typedef struct pinctrl_soc_pin_t {
riscv/riscv-privilege/sifive-freedom/pinctrl_soc.h:15:} pinctrl_soc_pin_t;
riscv/riscv-ite/common/pinctrl_soc.h:33:typedef struct pinctrl_soc_pin pinctrl_soc_pin_t;
xtensa/esp32s2/pinctrl_soc.h:28:} pinctrl_soc_pin_t;
xtensa/esp32/pinctrl_soc.h:28:} pinctrl_soc_pin_t;
arm/nuvoton_npcx/common/pinctrl_soc.h:125:typedef struct npcx_pinctrl pinctrl_soc_pin_t;
arm/rpi_pico/rp2/pinctrl_soc.h:34:typedef struct rpi_pinctrl_soc_pin pinctrl_soc_pin_t;
arm/renesas_smartbond/da1469x/pinctrl_soc.h:20:typedef struct smartbond_pinctrl_soc_pin pinctrl_soc_pin_t;
```

so, should the issue be fixed?

carlocaione reviewed (dismissed):
LGTM for the architectural part

xakep-amatop commented:
Hi, @aaillet
could you please re-review changes and answer the question ?

xakep-amatop commented:
fixed merging conflict

xakep-amatop commented:
fixed twister `tests/subsys/mgmt/mcumgr/smp_version` test for a new platforms

aaillet commented:
> Hi, @aaillet could you please re-review changes and answer the question ?

Working on another review !

Concerning questions, it's about maintainer ?
I talked with my internal Renesas contact and we are not sure yet of which team should maintain this part !
I don't think we have the time at Iot.bzh to maintain and test correctly arm64 part of these SoCs.
Discussions are in progress.

xakep-amatop commented:
> > Hi, @aaillet could you please re-review changes and answer the question ?
> 
> Working on a review !
> 
> Concerning questions, it's about maintainer ? I talked with my internal Renesas contact and we are not sure yet of which team should maintain this part ! I don't think we have the time at Iot.bzh to maintain and test correctly arm64 part of these SoCs. Discussions are in progress.

got it, thank you

lorc commented:
Hi @aaillet,

I can maintain this part of the code. Our company (EPAM Systems) is working closely with Renesas also.  I am already maintainer of Renesas platform in OP-TEE: https://github.com/OP-TEE/optee_os/blob/master/MAINTAINERS#L220 and collaborator of Xen-related stuff in Zephyr: https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml#L1461
 
So I am ready to act as maintainer/collaborator for GEN3/aarch64 part as well.

carlocaione reviewed (dismissed):
LGTM for the arch part

xakep-amatop commented:
Hi, @aaillet 

I would really apreciate if you could review my changes

lorc commented:
Hi @aaillet,

If you are okay with having me as maintainer, could you please point @xakep-amatop where he should put me in the `MAINTAINERS.yml`  ? Should I be in the `maintainers` or `collaborators` section?

aaillet commented:
> Hi @aaillet,
> 
> If you are okay with having me as maintainer, could you please point @xakep-amatop where he should put me in the `MAINTAINERS.yml` ? Should I be in the `maintainers` or `collaborators` section?

Hi, there is two possible solution:
- @xakep-amatop is creating a new "renesas r-car arm64" area to the maintainer file and add you as maintainer
- I create another PR when this one will be merged to do it my self

Creating a new "area" should be done alongside the add of a new github label for the bot to assign. The new maintainer will also be added to zephyr github teams to get corresponding rights on the project.



> Hi, @aaillet
> 
> I would really apreciate if you could review my changes

Sorry for the delay, I need to find time to do it asap, I'm starting to test new drivers version on our arm32 targets today to validate functioning.

nordicjm reviewed (dismissed):
MCUmgr change OK

aaillet commented:
I validated that these changes didn't broke R7 (arm32) support of H3.
I'm deeply reviewing clock control changes, I'll let you know.

Can you tell me what do you want to do concerning maintainer assignment ?

> Hi, there is two possible solution:
> 
> * @xakep-amatop is creating a new "renesas r-car arm64" area to the maintainer file and add you as maintainer
> * I create another PR when this one will be merged to do it my self (I have some update to my area to do, so I can do it at the same time)

xakep-amatop committed:
MAINTAINERS: Renesas R-Car ARM64: add entry

Renesas R-Car ARM64 platforms entry has been added to maintainers list.
lorc (Volodymyr Babchuk <volodymyr_babchuk@epam.com>) will be a maintainer
of the Renesas R-Car boards based on ARMv8.

Signed-off-by: Mykola Kvach <mykola_kvach@epam.com>
Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

xakep-amatop commented:
rebased onto the latest changes + added `Renesas R-Car ARM64` entry to the maintainers list ([last commit](https://github.com/zephyrproject-rtos/zephyr/pull/57172/commits/01f74f9eea8b08436a1f84a9ce4fbbc24d793b5e))

carlocaione reviewed (approved):
For the arch part

lorc reviewed (approved):
For adding me as a maintainer

de-nordic reviewed (approved):
MCUmgr is OK.

carlescufi commented:
Compliance failure is the usual pin control typedef.

aaillet commented:
In fact PR wasn't ready to be merged, I was working on the clock changes review as it is containing errors and kept my "change needed" while doing it.
Sorry I took that much time, I will then address these errors in a new PR !

carlocaione commented:
> In fact PR wasn't ready to be merged

Next time please keep it in draft state until ready to be merged.

aaillet commented:
> > In fact PR wasn't ready to be merged
> 
> Next time please keep it in draft state until ready to be merged.

I'm not PR author, was I able to do that ?
I wasn't aware, that's why I used "changes needed" to block the PR

xakep-amatop commented:
> > I validated that these changes didn't broke R7 (arm32) support of H3. I'm deeply reviewing clock control changes, I'll let you know.
> 
> I said that I was working on the clock control review (but not updated my status) as maintainer, but anyway, I'll fix it on another PR !

Feel free to write comments about any issues you've found in the CPG driver here. If you'd like, I can address them and open a new pull request.
